crypto/ec/ecp_secp384r1.c

```diff
@@ -0,0 +1,19653 @@
+/*-
+ * MIT License
+ *
+ * Copyright (c) 2020 Luis Rivera-Zamarripa, Jesús-Javier Chi-Domínguez, Billy
+ * Bob Brumley
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+#if defined(__SIZEOF_INT128__) && !defined(PEDANTIC)
+
+#include <stdint.h>
+#include <string.h>
+#define LIMB_BITS 64
+#define LIMB_CNT 6
+/* Field elements */
+typedef uint64_t fe_t[LIMB_CNT];
+typedef uint64_t limb_t;
+
+#define fe_copy(d, s) memcpy(d, s, sizeof(fe_t))
+#define fe_set_zero(d) memset(d, 0, sizeof(fe_t))
+
+/* Projective points */
+typedef struct {
+    fe_t X;
+    fe_t Y;
+    fe_t Z;
+} pt_prj_t;
+
+/* Affine points */
+typedef struct {
+    fe_t X;
+    fe_t Y;
+} pt_aff_t;
+
+/* BEGIN verbatim fiat code https://github.com/mit-plv/fiat-crypto */
+/*-
+ * MIT License
+ *
+ * Copyright (c) 2020 the fiat-crypto authors (see the AUTHORS file)
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
```

Comment:
@mattcaswell @levitte are there potential licensing issues to address here?

Reply:
My understanding is that the fiat code is under the MIT license. I think that if we can't get this under the Apache license, we need an OMC exception to allow this.

Reply:
There's some prior discussion on this at https://github.com/mit-plv/fiat-crypto/issues/678, in particular,
>  given the two institutions holding copyright on this project, and that any licensing changes at this point would be rather cumbersome and come at a significant cost of dev time better spent on improving the library.

So I think it's unlikely that we'll manage a license change at this point, though if it's important, we can look into it.  (If you want me to reopen https://github.com/mit-plv/fiat-crypto/issues/678, leave a comment there with your concerns and I can reopen it.)

Reply:
> My understanding is that the fiat code is under the MIT license. I think that if we can't get this under the Apache license, we need an OMC exception to allow this.

I agree. I think this is likely to be a significant problem. I don't recall any previous occasion where the OMC has granted such an exception. Another issue is that we require all original authors to provide a CLA. Do we actually know who all the original authors of this code are?

Reply:
I've asked the OMC about this. I'm not feeling very hopeful.

Reply:
MIT is not viral so I fail to see what the issue is. You can e.g. fork and relicense. Would you like help from our lawyers?

Reply:
There are two issues here:

1) CLAs. It is OpenSSL policy that *all* original authors of code must sign a CLA (except where a contribution is trivial in nature). So far we have been successful in applying this policy to all contributions that have occurred since 1.1.0 in 2016. So, under this policy, anyone who has had anything to do with this MIT licensed code at any point would have to sign the CLA.

Do we know who all these people are, and would they all be willing to sign a CLA?

If they all sign the CLA then we are actually at liberty to just remove the MIT licence entirely and re-license the code under our own Apache 2 licence (because in signing the CLA the authors allow us to do that).

This is not a perfect solution though. Since this is auto-generated code presumably we may wish to re-run that auto-generation at some point in the future (probably with some later version of the tool). Or if we wish to add support for new curves then again we would have to re-run it. In that case we have to check the history of the code again and get any new CLAs that are required. Potentially, even for contributors for whom we have a CLA, if they've made additional contributions to the code then we have to reconfirm with them that they are happy for their new contributions to be submitted under that CLA.

A possible way to side-step this issue is to get the OMC to grant an exception to the CLA rule for this code. While in theory this is possible it has *never* happened for any contribution since we introduced the CLA rule. It has on occasion been discussed for some PR or other but it has never been followed through to an actual vote. My expectation based on previous discussions is that the OMC would be reluctant to grant such an exception. I can't say for sure though unless an actual vote is held.

For the sake of argument lets say the OMC does grant an exception then that leads to issue 2:

2) Licence clash. You say above: "You can e.g. fork and relicense". The MIT license does not give us the ability to just apply a different licence instead. IIUC, we *can* apply an additional licence. So, in that case, both the Apache v2 *and* the MIT licence would apply to this code. While theoretically possible (they are not mutually incompatible) this effectively pollutes the OpenSSL licensing state. Until now the message is "OpenSSL is licensed under Apache v2". After this PR it would be "OpenSSL is licensed under Apache v2, except for this bit of code with is additionally licensed under MIT". I think that message is highly undesirable and confusing. Again I think we'd need an additional OMC exception to allow this approach.


Another option is to not commit the auto-generated bits at all. In that approach we would make it a requirement on the person building OpenSSL to ensure they have the relevant auto-generation tools available. Probably we would have to make it an optional dependency because 99.9% of people building OpenSSL won't have these tools readily available. But that of course seriously diminishes the value of the contribution because (probably) most people would simply skip building it.

Reply:
You already have many files that are dual licensed; "git grep CRYPTOGAMS" or look at crypto/LP* files.

Reply:
> You already have many files that are dual licensed;

Yes - those files are *dual* licensed under the normal sense of what that means (you get to choose which licence you want to use for those code files).

This is not *dual* licensing in the normal sense of the term. All the code in OpenSSL is available under the Apache v2 licence and only the Apache v2 licence if that's the one you want to use. If your lawyers just verify they are happy with Apache v2 then all the code is covered by that - end of story. It so happens that *some* code is *alternatively* available under a different licence. If you just want to cherry-pick those code files then you have the added option of using that code under different a different licence if you wish.

This is different. Its going back to the model where *both* licences apply. You can no longer make the statement that all code in OpenSSL is available under the Apache v2 licence. Some code files have *extra* requirements not stated in the Apache v2 licence.

Reply:
The files you point at are all offered explicitly under the OpenSSL license (Apache 2.0) and the authors have provided the required CLAs.

@richsalz you are actually mixing two different issues here - the code in question is only under a single license and it isn't the apache 2.0 license - that's the first issue; and separately CLAs serve a specific purpose beyond that of simple license compatibility - and that also isn't covered

Reply:
I didn't mix the issues; I explicitly did not talk about CLA. But I have a better understanding of the concern. You could put an openssl copyright on the code and make it dual-licensed if you wanted, or delve into the concept of "collective copyright" if you wanted.  I will stop contributing on this now.

Reply:
The [fiat-crypto `AUTHORS` file](https://github.com/mit-plv/fiat-crypto/blob/master/AUTHORS) lists the copyright holders:
Andres Erbsen
Google Inc.
Jade Philipoom
Massachusetts Institute of Technology
Zoe Paraskevopoulou

According to a brief chat with @davidben (and indirectly with @agl), we expect Google would be fine with dual-licensing the fiat-crypto code under user's-choice of MIT and Apache.  @andres-erbsen and @jadephilipoom are also happy with that, and I'm hopeful that @achlipala may be able to get MIT's permission to re-license/dual-license (though I don't know what's involved in that).  @zoep, would you also be fine with this relicensing from MIT to user's choice: MIT or Apache v2?

I'm not sure what to say about CLA's, that seems to be more complicated, especially with regard to future changes.

Reply:
Thanks @JasonGross for looking  in to this. Is that AUTHORS file the complete list of all authors who might have a copyright interest in the code (including minor but non-trivial contributions)? Is this a relatively stable list? I'm used to working on OpenSSL code which has hundreds of authors associated with it. This is a comparatively small list so perhaps the CLA issue is not so bad. We'd still need to get CLAs from all contributors (including future ones if we regenerate the code) - but if the list of authors is small then that could  be do-able. There would of course be a risk that we find ourselves stuck if some future contributor to fiat-crypto declines to sign *our* CLA. Suddenly at that point we would be unable to regenerate the code any more with the latest versions of fiat-crypto.

Reply:
The complete list of contributors is:
```
$ git shortlog -nse
  3928  Jason Gross <jgross@mit.edu>
  1066  Jade Philipoom <jadep@mit.edu>
   609  Andres Erbsen <andreser@mit.edu>
   254  Robert Sloan <varomodt@gmail.com>
    21  Adam Chlipala <adamc@csail.mit.edu>
     9  David Benjamin <davidben@google.com>
     9  Zoe Paraskevopoulou <zoopar@fb.com>
     7  François Garillot <francois@garillot.net>
     6  Hugo Herbelin <Hugo.Herbelin@inria.fr>
     5  Benjamin Barenblat <bbaren@google.com>
     5  Maxime Dénès <maxime.denes@inria.fr>
     4  Samuel Gruetter <samuel.gruetter@epfl.ch>
     3  Frédéric Besson <frederic.besson@inria.fr>
     3  Jasper Hugunin <jasperh@cs.washington.edu>
     2  Automated Publisher <actions@users.noreply.github.com>
     2  Vincent Laporte <Vincent.Laporte@inria.fr>
     1  Daniel Hirche <daniel@xcx.cc>
     1  George Tankersley <george.tankersley@gmail.com>
     1  Pierre Letouzey <pierre.letouzey@inria.fr>
```
People are supposed to update `AUTHORS` when contributing if they care about retaining copyright, but it's possible some contributors do care but didn't realize they would do this.  Idk what the legal status is of PR without adding themselves to `AUTHORS`.

In any case, the first 7 of these entries are either in AUTHORS or covered by an institution in the file.

Regarding the others:
- François Garillot (`@huitseeker`) made minor edits to the Rust generation code and set up the Rust Crate packaging.  He didn't touch anything in the pipeline generating C code
- Hugo Herbelin, Maxime Dénès, Frédéric Besson, Vincent Laporte, Jasper Hugunin, and Pierre Letouzey are all Coq devs who I believe have contributed only minor patches to make fiat-crypto compatible with some change in Coq
- Benjamin Barenblat, I believe, is covered by the Google AUTHOR entry
- Samuel Gruetter, I believe, is covered by the MIT AUTHOR entry
- Daniel Hirche fixed a typo in the naming of one of the primes (we had committed the code for p434 under the name p484)
- George Tankersley fixed a typo in the readme where we had the wrong path to a script

I don't believe any of these qualify for copyright holdership, though IANAL.

So far, the AUTHORS list has been relatively stable, gaining a new entry perhaps once a year (in the relatively short lifetime of this project).

Reply:
I've gotten confirmation from @zoep that she's fine with us relicensing, so if it'd help you, we can relicense to be dual-licensed, user's choice, MIT or Apache 2.0.  However, my current understanding is that MIT doesn't permit us to sign CLA's, so I don't think you'll be able to get CLAs for fiat-crypto.  Given this, would you like us to relicense fiat-crypto to allow use under Apache?

Reply:
Wow! @JasonGross - that's fantastic that you're willing to relicense!! What an amazing offer. However without a solution to the CLA problem we still have an issue. Are you sure about the MIT policy? I notice several mit.edu addresses in our CLA database. Presumably you individually own the copyright on your own submissions (not MIT as an institution), so these CLAs would be individual agreements?

Reply:
Sorry for the delay in getting back.  I'm not sure what's going on in the other cases; in our case, MIT owns the copyright, and this is MIT's policy on code where it owns copyright.

Reply:
Some people on the Go issue tracker (https://github.com/golang/go/issues/40171#issuecomment-694629043) suggested:

> An alternative solution to our problem of imposing new notice requirements on every Go binary would be if the generator _outputs_ could be licensed under a non-attribution license such as [MIT-0](https://spdx.org/licenses/MIT-0.html) or a source-code-attribution-only license such as [BSD-1-Clause](https://spdx.org/licenses/BSD-1-Clause.html).

Would either of these licenses allow integration of the generated code into OpenSSL without us needing to sign CLAs?

Reply:
Well there are two  issues: licence and CLAs. Your offer above to relicence to Apache 2.0 solves the licence issue.  The CLA issue is independent of licensing. So, unfortunately, I don't think swapping to a different licence solves it.